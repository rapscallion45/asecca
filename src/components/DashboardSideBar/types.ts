/*
 * Dashboard Sidebar type definitions
 *
 * These type definitions are used in the applications Dashboard Sidebar component.
 */

/**
 * Nav Bar sidebar on close callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback IDashboardSidebarOnCloseCallback
 * @returns {void} - no return value
 */
export interface IDashboardSidebarOnCloseCallback {
  (): void;
}
