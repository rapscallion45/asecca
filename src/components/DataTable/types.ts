/*
 * Data Table type definitions
 *
 * These type definitions are used in the applications data table component.
 * Only make changes to this file if the corresponding change is made in the
 * Data Table component(s)
 */

/**
 * Data Table Row Cell Value types
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof DataTable
 *
 * @typedef DataTableRowCellValue
 */
export type DataTableRowCellValue =
  | string
  | Array<string>
  | boolean
  | null
  | undefined;

/**
 * Data Table Column types
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 * @memberof DataTable
 *
 * @typedef DataTableColumnType
 */
export type DataTableColumnType = 'string' | 'currency' | 'select' | 'check';

/**
 * Data Table Column properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IDataTableColumn
 * @prop {string} label - data table column name
 * @prop {string} key - data table column data identifier
 * @prop {DataTableColumnType} type - data table column type
 */
export interface IDataTableColumn {
  label: string;
  key: string;
  type: DataTableColumnType;
}

/**
 * Data Table Row properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IDataTableRow
 * @prop {string} label - data table row name
 */
export interface IDataTableRow {
  label: string;
}

/**
 * Data Table Cell Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableEditCellValueCallback
 * @param {DataTableRowCellValue} value - updated cell value, can be null
 * @param {string} colKey - cell column data identifier
 * @param {number} rowIdx - cell row index in data table
 * @returns {void} - no return value
 */
export interface IDataTableEditCellValueCallback {
  (value: DataTableRowCellValue, colKey: string, rowIdx: number): void;
}

/**
 * Data Table Cell Value Retrieval callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableGetCellValueCallback
 * @param {number} rowIdx - cell row index in data table
 * @param {IDataTableColumn} column - cell column in data table
 * @returns {DataTableRowCellValue} - cell value, can be null or undefined
 */
export interface IDataTableGetCellValueCallback {
  (rowIdx: number, column: IDataTableColumn): DataTableRowCellValue;
}

/**
 * Data Table Currency Cell Value Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableEditCurrencyCellValueCallback
 * @param {string | null} value - updated cell value
 * @returns {void} - no return value
 */
export interface IDataTableEditCurrencyCellValueCallback {
  (value: string | null): void;
}

/**
 * Data Table Checkbox Cell Value Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 *
 * @callback IDataTableEditCheckboxCellValueCallback
 * @param {boolean | null} value - updated cell boolean value
 * @returns {void} - no return value
 */
export interface IDataTableEditCheckboxCellValueCallback {
  (value: boolean | null): void;
}
