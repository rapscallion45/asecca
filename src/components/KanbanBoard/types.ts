/*
 * Kanban Board type definitions
 *
 * These type definitions are used in the applications kanban board component.
 * Only make changes to this file if the corresponding change is made in the
 * Kanban Board component(s)
 */

/**
 * Kanban Board Column On Drop callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.7
 *
 * @callback IKanbanBoardColumnOnDropCallback
 * @param {number} colIndex - column index of task
 * @param {number} prevColIndex - previous column index of task
 * @param {number} taskIndex - index of task
 * @returns {void} - no return value
 */
export interface IKanbanBoardColumnOnDropCallback {
  (colIndex: number, prevColIndex: number, taskIndex: number): void;
}
