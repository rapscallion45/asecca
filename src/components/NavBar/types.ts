/*
 * Nav Bar type definitions
 *
 * These type definitions are used in the applications Nav Bar component.
 */

/**
 * Nav Bar sidebar open callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback INavBarOnSidebarOpenCallback
 * @returns {void} - no return value
 */
export interface INavBarOnSidebarOpenCallback {
  (): void;
}
