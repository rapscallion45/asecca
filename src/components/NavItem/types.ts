/*
 * Nav Item type definitions
 *
 * These type definitions are used in the applications Nav Item component.
 */

/**
 * Nav Item configuration properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof NavItem
 *
 * @typedef INavItemConfig
 * @prop {string} title - nav item title text
 * @prop {string} path - nav item URL path
 * @prop {any} icon - nav item icon
 * @prop {boolean} newTab - nav item link opens in new tab
 * @prop {Array<INavItemConfig>} children - nav item child components
 */
export interface INavItemConfig {
  title: string;
  path?: string;
  icon?: any;
  newTab?: boolean;
  children?: Array<INavItemConfig>;
}
