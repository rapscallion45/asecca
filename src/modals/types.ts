/**
 * Type and interface definitions for global application modals
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 */

import { SvgIconProps } from '@mui/material';

/**
 * Modal button size options
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof Modals
 *
 * @typedef ModalButtonSizeType - can either be 'small', 'medium' or 'large'
 */
export type ModalButtonSizeType = 'small' | 'medium' | 'large';

/**
 * Modal button icon size options
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof Modals
 *
 * @typedef ModalButtonIconSizeType - can either be 'small', 'inherit' or 'large'
 */
export type ModalButtonIconSizeType = 'small' | 'inherit' | 'large';

/**
 * Modal button colour options
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof Modals
 *
 * @typedef ModalButtonColourType - can either be 'primary', 'secondary', 'inherit' or 'error'
 */
export type ModalButtonColourType =
  | 'primary'
  | 'secondary'
  | 'inherit'
  | 'error';

/**
 * Modal button variant options
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof Modals
 *
 * @typedef ModalButtonVariantType - can either be 'outlined', 'contained' or 'text'
 */
export type ModalButtonVariantType = 'outlined' | 'contained' | 'text';

/**
 * Modal normal button type definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef INormalButtonType
 * @prop {string} type - type of button
 * @prop {string} text - display text of button
 * @prop {string} icon - button icon element
 * @prop {ModalButtonSizeType} size - button size option
 * @prop {string} style - saving state flag of costs config data
 * @prop {ModalButtonVariantType} variant - MUI button variant
 * @prop {string} className - MUI icon button class
 * @prop {ModalButtonColourType} color - button color option
 */
export interface INormalButtonType {
  type: 'normal';
  text?: string;
  icon: (props: SvgIconProps) => JSX.Element;
  size?: ModalButtonSizeType;
  style?: { [name: string]: string } | { [name: string]: number };
  variant?: ModalButtonVariantType;
  className?: string;
  color?: ModalButtonColourType;
}

/**
 * Modal round button type definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IRoundIconButtonType
 * @prop {string} type - type of button
 * @prop {string} icon - button icon element
 * @prop {ModalButtonSizeType} size - button size option
 * @prop {string} style - saving state flag of costs config data
 * @prop {ModalButtonVariantType} variant - MUI button variant
 * @prop {ModalButtonColourType} color - button color option
 */
export interface IRoundIconButtonType {
  type: 'round';
  icon: (props: SvgIconProps) => JSX.Element;
  size?: 'small' | 'medium' | 'large';
  style?: { [name: string]: string } | { [name: string]: number };
  variant?: 'outlined' | 'contained' | 'text';
  color?: 'primary' | 'secondary' | 'inherit';
}

/**
 * Modal icon button type definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IIconButtonType
 * @prop {string} type - type of button
 * @prop {string} icon - button icon element
 * @prop {ModalButtonSizeType} size - button size option
 * @prop {string} style - saving state flag of costs config data
 * @prop {ModalButtonIconSizeType} iconSize - button icon size option
 * @prop {string} className - MUI icon button class
 * @prop {ModalButtonColourType} color - button color option
 */
export interface IIconButtonType {
  type: 'icon';
  icon: (props: SvgIconProps) => JSX.Element;
  size?: ModalButtonSizeType;
  style?: { [name: string]: string } | { [name: string]: number };
  iconSize?: 'small' | 'inherit' | 'large';
  className?: string;
  color?: 'primary' | 'secondary' | 'inherit';
}

/**
 * Modal menu button type definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IMenuItemButtonType
 * @prop {string} type - type of button
 * @prop {string} text - display text of button
 * @prop {string} icon - button icon element
 * @prop {string} closeMenu - callback handler for closing the modal
 * @prop {string} iconStyle - icon style options
 * @prop {string} className - MUI icon button class
 * @prop {ModalButtonColourType} color - button color option
 */
export interface IMenuItemButtonType {
  type: 'menu';
  text: string;
  icon: (props: SvgIconProps) => JSX.Element;
  closeMenu: () => void;
  iconStyle?: { [name: string]: string } | { [name: string]: number };
  className?: string;
  color?: 'primary' | 'secondary' | 'inherit' | 'error';
}

/**
 * Modal menu button type definition
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IFabButtonType
 * @prop {string} type - type of button
 * @prop {string} icon - button icon element
 * @prop {ModalButtonSizeType} size - button size option
 * @prop {ModalButtonVariantType} variant - MUI button variant
 * @prop {string} text - display text of button
 * @prop {ModalButtonColourType} color - button color option
 */
export interface IFabButtonType {
  type: 'fab';
  icon: (props: SvgIconProps) => JSX.Element;
  size?: 'small' | 'medium' | 'large';
  variant?: 'extended' | 'circular';
  text?: string;
  color?: 'primary' | 'secondary' | 'inherit';
}

/**
 * Modal trigger button types
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof Modals
 *
 * @typedef TriggerButtonTypes - trigger button options type
 */
export type TriggerButtonTypes =
  | INormalButtonType
  | IRoundIconButtonType
  | IIconButtonType
  | IMenuItemButtonType
  | IFabButtonType;
