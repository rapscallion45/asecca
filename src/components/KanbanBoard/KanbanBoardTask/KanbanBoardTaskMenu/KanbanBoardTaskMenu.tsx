import { FC, useState, useCallback, MouseEvent } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import DeleteOutlineIcon from '@mui/icons-material/DeleteOutline';
import EditOutlinedIcon from '@mui/icons-material/EditOutlined';
import MoreVertIcon from '@mui/icons-material/MoreVert';
import { alpha } from '@mui/material/styles';
import { Button, Box, IconButton } from '@mui/material';
import { AppState } from '@/redux/store';
import { IEditKanbanBoardTaskPayload } from '@/redux/types';
import { deleteTask } from '@/redux/slices/kanbanSlice';
import MenuPopover from '@/components/MenuPopover/MenuPopover';
import ConfirmDialog from '@/modals/ConfirmModal/ConfirmModal';
import FormDialog from '@/modals/FormModal/FormModal';
import { ModalButtonIconSizeType } from '@/modals/types';
import { IKanbanBoard } from '@/lib/api/api-types';
import KanbanBoardTaskForm from '../KanbanBoardTaskForm/KanbanBoardTaskForm';

/**
 * Kanban Board Task Menu Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.2
 *
 * @typedef IKanbanBoardTaskMenuProps
 * @prop {number} colIndex - column index of this task
 * @prop {number} taskIndex - index of task
 * @prop {IEditKanbanBoardTaskPayload} currentData - current task data
 * @prop {ModalButtonIconSizeType} iconSize - button icon size
 */
interface IKanbanBoardTaskMenuProps {
  colIndex: number;
  taskIndex: number;
  currentData: IEditKanbanBoardTaskPayload;
  iconSize?: ModalButtonIconSizeType;
}

/**
 * Kanban Board Task Menu
 *
 * Kanbaord board task menu interface
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.2
 *
 * @component
 * @param {IKanbanBoardTaskMenuProps} props - component props
 * @returns {FC} - kanban board task menu functional component
 */
const KanbanBoardTaskMenu: FC<IKanbanBoardTaskMenuProps> = (props) => {
  const { colIndex, taskIndex, currentData, iconSize } = props;
  const dispatch = useDispatch();
  const { data: kanbanData } = useSelector((state: AppState) => state.kanban);
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);

  /**
   * Callback listener for opening task menu popover
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.2
   *
   * @method
   * @param {MouseEvent<HTMLButtonElement>} event - object change event
   */
  const handleOpenMenu = useCallback((event: MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  }, []);

  /**
   * Callback listener for closing task menu popover
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.2
   *
   * @method
   */
  const handleCloseMenu = useCallback(() => {
    setAnchorEl(null);
  }, []);

  /**
   * Callback listener for deleting task
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.2
   *
   * @method
   * @param {number} deleteColIndex - column of task to be deleted
   * @param {number} deleteTaskIndex - index of task to be deleted
   * @param {any} closeModal - callback for handling closing of modal
   */
  const handleDelete = (
    deleteColIndex: number,
    deleteTaskIndex: number,
    closeModal: () => void
  ) => {
    dispatch(
      deleteTask({ colIndex: deleteColIndex, taskIndex: deleteTaskIndex })
    );
    closeModal();
  };

  return (
    <>
      <IconButton
        size="small"
        color={anchorEl ? 'primary' : 'default'}
        onClick={handleOpenMenu}
        sx={{
          ...(Boolean(anchorEl) && {
            bgcolor: (theme) =>
              alpha(
                theme.palette.primary.main,
                theme.palette.action.focusOpacity
              ),
          }),
        }}
      >
        <MoreVertIcon fontSize={iconSize || 'medium'} />
      </IconButton>
      <MenuPopover
        open={Boolean(anchorEl)}
        onClose={handleCloseMenu}
        keepMounted
        anchorEl={anchorEl}
        sx={{ width: 220 }}
      >
        <FormDialog
          triggerBtn={{
            type: 'menu',
            // @ts-ignore
            icon: EditOutlinedIcon,
            iconStyle: { marginRight: '10px' },
            text: 'Edit Task Info',
            closeMenu: handleCloseMenu,
          }}
          title="Edit Task Info"
        >
          <KanbanBoardTaskForm
            isEditMode
            columns={
              kanbanData.boards.find((item: IKanbanBoard) => item.isActive)
                ?.columns || []
            }
            currentData={currentData}
            closeModal={handleCloseMenu}
          />
        </FormDialog>
        <ConfirmDialog
          title="Confirm Delete Task"
          contentText="Are you sure you want to permanently delete this task?"
          actionBtnText="Delete"
          triggerBtn={{
            type: 'menu',
            text: 'Delete Task',
            // @ts-ignore
            icon: DeleteOutlineIcon,
            iconStyle: { marginRight: '10px' },
            closeMenu: handleCloseMenu,
          }}
          // processing={deleting}
          actionFunc={(closeDialog) =>
            handleDelete(colIndex, taskIndex, closeDialog)
          }
        />

        <Box sx={{ p: 2, pt: 1.5 }}>
          <Button
            fullWidth
            color="inherit"
            variant="outlined"
            onClick={handleCloseMenu}
          >
            Close
          </Button>
        </Box>
      </MenuPopover>
    </>
  );
};
export default KanbanBoardTaskMenu;
