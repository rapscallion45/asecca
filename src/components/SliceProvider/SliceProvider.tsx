import { FC, ReactNode, createContext, useContext } from 'react';
import { useSelector } from 'react-redux';
import { Slice } from '@reduxjs/toolkit';
import { AppState } from '@/redux/store';

/**
 * Default value for the slice context
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof SliceProvider
 *
 * @constant
 * @type {Slice<Object>}
 */
const SliceContext = createContext<Slice>({} as Slice);

/**
 * Slice Provider Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 *
 * @typedef ISliceProviderProps
 * @prop {Slice} slice - context slice to provide down to child nodes
 * @prop {ReactNode} children - child nodes
 */
interface ISliceProviderProps {
  slice: Slice;
  children: ReactNode;
}

/**
 * Slice Provider
 *
 * Slice provider functional component wrapper for providing global state
 * managment functionality to child components
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 *
 * @component
 * @param {ISliceProviderProps} props - component props
 * @returns {FC} - slice provider functional component
 */
const SliceProvider: FC<ISliceProviderProps> = (props) => {
  const { slice, children } = props;
  return (
    <SliceContext.Provider value={slice}>{children}</SliceContext.Provider>
  );
};

/**
 * Provider specific actions interface for passed slice
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof KanbanReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {CaseReducerActions<any, string>}
 */
export const useSliceActions = () => useContext<Slice>(SliceContext).actions;

/**
 * Provider specific selector interface for passed slice
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof KanbanReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {AppState}
 */
export const useSliceSelector = () => {
  const { name } = useContext<Slice>(SliceContext);
  return useSelector((state: AppState) => state[name as keyof AppState]);
};

export default SliceProvider;
