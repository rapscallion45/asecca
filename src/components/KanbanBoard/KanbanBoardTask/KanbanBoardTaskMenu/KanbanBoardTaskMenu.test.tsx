import React from 'react';
import { Provider } from 'react-redux';
import {
  render,
  screen,
  waitFor,
  act,
  fireEvent,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import {
  collectionsKanbanSlice,
  fetchByProjectId as fetchKanbanBoardByProjectId,
} from '@/redux/slices/collectionsKanbanSlice';
import store from '@/redux/store';
import SliceProvider from '@/components/SliceProvider/SliceProvider';
import { IEditKanbanBoardTaskPayload } from '@/redux/types';
import KanbanBoardTaskMenu from './KanbanBoardTaskMenu';

/* default test query ID */
const query: string = '66135000001760012';

/* test index */
const testIdx: number = 0;

/* test task data */
const testTaskData: IEditKanbanBoardTaskPayload = {
  name: 'Test Task',
  status: 'Todo',
  newColIndex: testIdx,
  taskIndex: testIdx,
  prevColIndex: testIdx,
};

/**
 * Kanban Board Task Menu Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Kanban Board Task Menu', () => {
  describe('Values', () => {
    /** ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should render menu icon button', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      render(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoardTaskMenu
              editData={testTaskData}
              colIndex={testIdx}
              taskIndex={testIdx}
            />
          </SliceProvider>
        </Provider>
      );

      /** Assert - correct value should be displayed from test data */
      await waitFor(() => {
        expect(screen.getByRole('button')).toBeInTheDocument();
      });
      expect(screen.getByTestId('MoreVertIcon')).toBeInTheDocument();
    });
  });

  describe('Functions', () => {
    /** ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should open and close menu popup on button clicks', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      render(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoardTaskMenu
              editData={testTaskData}
              colIndex={testIdx}
              taskIndex={testIdx}
            />
          </SliceProvider>
        </Provider>
      );

      /** Assert - menu button should be rendered and popup not displayed */
      await waitFor(() => {
        expect(screen.getByRole('button')).toBeInTheDocument();
      });
      expect(screen.getByTestId('MoreVertIcon')).toBeInTheDocument();
      expect(screen.queryByRole('presentation')).toBeNull();

      /** Act - click on menu button to open menu popup */
      act(() => {
        fireEvent(
          screen.getByTestId('MoreVertIcon'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });

      /** Assert - ensure menu popup is displayed */
      await waitFor(() => {
        expect(screen.getByRole('presentation')).toBeInTheDocument();
      });
      expect(screen.getByText('Edit Task Info')).toBeInTheDocument();
      expect(screen.getByText('Close')).toBeInTheDocument();

      /** Act - click on Close button to close menu popup */
      act(() => {
        fireEvent(
          screen.getByText('Close'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });

      /** Assert - ensure menu popup is closed */
      await waitFor(() => {
        expect(screen.queryByRole('presentation')).toBeNull();
      });
    });

    // it('Should delete task if delete button clicked', async () => {
    //   /** Arrange */
    //   act(() => {
    //     store.dispatch(
    //       fetchKanbanBoardByProjectId({
    //         projectId: query,
    //       })
    //     );
    //   });

    //   /** Act */
    //   render(
    //     <Provider store={store}>
    //       <SliceProvider slice={collectionsKanbanSlice}>
    //         <KanbanBoardTaskMenu
    //           currentData={testTaskData}
    //           colIndex={testIdx}
    //           taskIndex={testIdx}
    //           canEdit
    //         />
    //       </SliceProvider>
    //     </Provider>
    //   );

    //   /** Assert - menu button should be rendered and popup not displayed */
    //   await waitFor(() => {
    //     expect(screen.getByRole('button')).toBeInTheDocument();
    //   });
    //   expect(screen.getByTestId('MoreVertIcon')).toBeInTheDocument();
    //   expect(screen.queryByRole('presentation')).toBeNull();

    //   /** Act - click on menu button to open menu popup */
    //   act(() => {
    //     fireEvent(
    //       screen.getByTestId('MoreVertIcon'),
    //       new MouseEvent('click', {
    //         bubbles: true,
    //         cancelable: true,
    //       })
    //     );
    //   });

    //   /** Assert - ensure menu popup is displayed */
    //   await waitFor(() => {
    //     expect(screen.getByRole('presentation')).toBeInTheDocument();
    //   });
    //   expect(screen.getByText('Delete Task')).toBeInTheDocument();
    //   expect(screen.getByText('Close')).toBeInTheDocument();

    //   /** Act - click on Close button to close menu popup */
    //   act(() => {
    //     fireEvent(
    //       screen.getByText('Delete Task'),
    //       new MouseEvent('click', {
    //         bubbles: true,
    //         cancelable: true,
    //       })
    //     );
    //   });

    //   /** Assert - ensure menu popup is closed */
    //   await waitFor(() => {
    //     expect(screen.queryByRole('presentation')).toBeNull();
    //   });
    // });
  });

  it('Renders correctly', async () => {
    /** Arrange */
    act(() => {
      store.dispatch(
        fetchKanbanBoardByProjectId({
          projectId: query,
        })
      );
    });

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoardTaskMenu
              editData={testTaskData}
              colIndex={testIdx}
              taskIndex={testIdx}
            />
          </SliceProvider>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
