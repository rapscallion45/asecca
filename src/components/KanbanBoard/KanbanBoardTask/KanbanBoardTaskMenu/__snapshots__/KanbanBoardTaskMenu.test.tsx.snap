// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kanban Board Task Menu Renders correctly 1`] = `
<button
  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
    data-testid="MoreVertIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
    />
  </svg>
</button>
`;
